// @microgamma/loggator/lib/lib/log.decorator.js
"use strict";
// tslint:disable: no-object-mutation no-string-literal no-parameter-reassignment
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const debug_1 = __importDefault(require("debug"));
require("reflect-metadata");
exports.debuggerMetadata = Symbol('debug');
let namespace;
const d = debug_1.default;
function Log(subspace) {
    return (target, propertyKey) => {
        Reflect.defineMetadata(exports.debuggerMetadata, subspace || target.constructor.name, target);
        target[propertyKey] = {
            // @ts-ignore
            d: (...args) => {
                subspace = subspace || target.constructor.name;
                // @ts-ignore
                const ns = namespace ? [namespace, subspace].join(':') : subspace;
                const loggerFn = getDebugger(ns);
                // @ts-ignore
                return loggerFn.apply(null, args);
            }
        };
    };
}
exports.Log = Log;
function setNamespace(ns) {
    namespace = ns;
}
exports.setNamespace = setNamespace;
function getDebugger(ns) {
    const de = d(ns);
    de['useColors'] = true;
    return de;
}
exports.getDebugger = getDebugger;
